swarm-derive/src/lib.rs

```diff
@@ -672,80 +672,47 @@ fn build_struct(ast: &DeriveInput, data_struct: &DataStruct) -> syn::Result<Toke
+                let event_variant: syn::Variant =
+                    syn::parse_str(&field.to_string().to_upper_camel_case())
+                        .expect("uppercased field name to be a valid enum variant name");
+                quote! { #out_event_name::#event_variant }
             } else {
-                quote! { event.into() }
+                quote! { |e| e.into() }
             };
 
-            quote! {
-                std::task::Poll::Ready(#network_behaviour_action::GenerateEvent(event)) => {
-                    return std::task::Poll::Ready(#network_behaviour_action::GenerateEvent(#into_out_event))
-                }
-            }
-        };
+            let map_in_event = quote! { |event| #wrapped_event };
 
-        quote!{
-            match #trait_to_impl::poll(&mut self.#field, cx) {
-                #generate_event_match_arm
-                std::task::Poll::Ready(#network_behaviour_action::Dial { opts }) => {
-                    return std::task::Poll::Ready(#network_behaviour_action::Dial { opts });
-                }
-                std::task::Poll::Ready(#network_behaviour_action::ListenOn { opts }) => {
-                    return std::task::Poll::Ready(#network_behaviour_action::ListenOn { opts });
-                }
-                std::task::Poll::Ready(#network_behaviour_action::RemoveListener { id }) => {
-                    return std::task::Poll::Ready(#network_behaviour_action::RemoveListener { id });
-                }
-                std::task::Poll::Ready(#network_behaviour_action::NotifyHandler { peer_id, handler, event }) => {
-                    return std::task::Poll::Ready(#network_behaviour_action::NotifyHandler {
-                        peer_id,
-                        handler,
-                        event: #wrapped_event,
-                    });
-                }
-                std::task::Poll::Ready(#network_behaviour_action::NewExternalAddrCandidate(addr)) => {
-                    return std::task::Poll::Ready(#network_behaviour_action::NewExternalAddrCandidate(addr));
-                }
-                std::task::Poll::Ready(#network_behaviour_action::ExternalAddrConfirmed(addr)) => {
-                    return std::task::Poll::Ready(#network_behaviour_action::ExternalAddrConfirmed(addr));
-                }
-                std::task::Poll::Ready(#network_behaviour_action::ExternalAddrExpired(addr)) => {
-                    return std::task::Poll::Ready(#network_behaviour_action::ExternalAddrExpired(addr));
-                }
-                std::task::Poll::Ready(#network_behaviour_action::CloseConnection { peer_id, connection }) => {
-                    return std::task::Poll::Ready(#network_behaviour_action::CloseConnection { peer_id, connection });
+            quote! {
+                match #trait_to_impl::poll(&mut self.#field, cx) {
+                    std::task::Poll::Ready(e) => return std::task::Poll::Ready(e.map_out(#map_out_event).map_in(#map_in_event)),
```

Comment:
@dhuseby The main idea is that we can delegate into `libp2p-swarm` here to map the individual variants for us which will enforce an exhaustive matching.

Reply:
that's a lot simpler. this is slick.